### ISSUE TITLE
What additional text is needed regarding non-compliant traffic marked as both NQB and ECT(1)?

### ISSUE THREAD

**Issue by gwhiteCL** [gwhiteCL/NQBdraft#33]
https://mailarchive.ietf.org/arch/msg/tsvwg/2jzEaHmGg-qSqLGTX_YbEyL4F5c/
From RG:
> I'm referring to the app programmer pursuing the biggest advantage for his app by being standards compliant where that is advantageous, i.e. an app-programmer cherry-picking features from L4S and NQB (including TCP prague requs) to optimize app throughput without obviously ignoring both standards simultaneously. To me, a standards track document must offer text how an access provider reliably fends off traffic of this kind (working detection and working mitigation). That latter part is at least incomplete and I don't think it addresses traffic marked DSCP NQB and ECT (1).

**Comment by gwhiteCL** [comment]
The text on this in draft-17 reads:

> Packets marked with both the NQB DSCP and the ECT(1) codepoint SHOULD be treated the same as packets marked with either codepoint alone by the traffic protection function (defined in [Section 5.2]) and by any re-marking/traffic policing function designed to protect unmanaged networks (as described in [Section 4.4.1]).

This is problematic in the case that there is any difference in treatment of NQB and ECT(1). For example, a traffic protection function for ECT(1) might try to validate that the microflow is complying with the Prague requirements (perhaps by observing the fraction of packets marked as CE).  Whereas a traffic protection function for NQB might try to validate that the microflow's burst rate remains beneath the rate equation in 4.1. 

In 58365e30b5a435245c1b6992521aa6c6a769c6b0 I've split the sentence into two sentences to address this:  

>  In nodes that support [RFC9330] and [RFC9331], packets marked with both the NQB DSCP and the ECT(1) codepoint SHOULD be treated the same as packets marked with ECT(1) alone by the L4S Network Node functions (Section 5 of [RFC9331]), and by any mechanisms designed to protect the L4S queue (such as those discussed in Section 8.2 of [RFC9330]). Packets marked with both the NQB DSCP and the ECT(1) codepoint SHOULD be treated the same as packets marked with NQB alone by the traffic protection function (defined in Section 5.2) and by any re-marking/traffic policing function designed to protect unmanaged networks (as described in Section 4.4.1).

If there is a preference for these to be MUSTs, I would not object to that.

**Comment by bbriscoe** [comment]
The intent of the new requirements is good - i.e. if mark with both, has to pass both tests.

I don't see any problem with saying MUST instead of SHOULD, because "...MUST be treated the same as..." doesn't make policing any more or less mandatory than it already is. Also, if the operator knows for sure that one policing function is stricter than the other, they still only have to apply the stricter function to comply with these MUSTs.

The wording is rather hard to follow. It might be better if each sentence is turned round so that the traffic protection functions are the subject of each sentence. But that might not improve the first sentence, which has a complex set of possible things as its subject. Another alternative would be to start with a briefer sentence like:

"...packets marked with both the NQB DSCP and the ECT(1) codepoint MUST be treated the same as NQB packets by NQB Network Node functions, and the same as ECT(1) packets by L4S Network Node functions. Here, NQB network node functions means..., and L4S Network Node functions means..."

**Comment by gwhiteCL** [comment]
Thanks Bob.  I received an off list comment that to me looks like the SHOULD might be preferred by some. 

On the wording, I tried out a few variants of your suggestions and realized that we've not taken into account an aspect that makes the language even more cumbersome.

The requirement statement on "L4S Network Node functions" is (in-effect) that they treat ECT(1) packets uniformly, regardless of DSCP.  This conflicts with some of the text in https://www.rfc-editor.org/rfc/rfc9331#section-5.4.1.3 and the concepts in https://datatracker.ietf.org/doc/html/draft-briscoe-tsvwg-l4s-diffserv-02.

I think we need to be more clear that an L4S node treats ECT(1) + NQB packets the same as ECT(1) + DSCP 0 packets.  It does not need to treat them the same as ECT(1) packets that are AF42 (for example).

**Comment by gwhiteCL** [comment]
Here's an attempt (maybe this isn't too cumbersome in the end):

> NQB network functions SHOULD treat packets marked with the NQB DSCP uniformly, regardless of the value of the ECN field. Here, NQB network functions means the traffic protection function (defined in Section 5.2) and any re-marking/traffic policing function designed to protect unmanaged networks (as described in Section 4.4.1). 

> L4S network functions SHOULD treat packets marked with the NQB DSCP and ECT(1) the same as packets marked with the Default DSCP and ECT(1). Here, L4S network functions means the L4S Network Node functions (Section 5 of [RFC9331]), and any mechanisms designed to protect the L4S queue (such as those discussed in Section 8.2 of [RFC9330]).


I think I'll go with this for now since the draft-18 deadline is approaching.

**Comment by gwhiteCL** [comment]
Maybe this should instead be:

> L4S network functions SHOULD treat packets marked with the NQB DSCP and ECT(1) or CE the same as packets marked with the Default DSCP and the same ECN value. Here, L4S network functions means the L4S Network Node functions (Section 5 of [RFC9331]), and any mechanisms designed to protect the L4S queue (such as those discussed in Section 8.2 of [RFC9330]).

What about NQB-marked ECT(0) and NQB-marked Not-ECT packets in an L4S network node?  Do we leave that unstated?

Do we need to make a general ECN recommendation?

RFC3168 already says:
>  The above discussion of when CE may be set instead of dropping a
>    packet applies by default to all Differentiated Services Per-Hop
>    Behaviors (PHBs) [RFC 2475].  Specifications for PHBs MAY provide
>    more specifics on how a compliant implementation is to choose between
>    setting CE and dropping a packet, but this is NOT REQUIRED.

**Comment by gorryfair** [comment]
Without reading the entire section, I'd say this looks about correct for L4S.
But then, is it correct also to include classical ECN, like this or not:

```
L4S network functions SHOULD treat packets marked with the NQB DSCP  and a non-zero ECN field the same as packets marked with the Default DSCP and the same ECN value. Here, L4S network functions refers to the L4S Network Node functions (Section 5 of [RFC9331]), and any mechanisms designed to protect the L4S queue (such as those discussed in Section 8.2 of [RFC9330]).
...
i.e. if packets pass the NQB test, then put them in the L4S queue?... or does this result in L4S marking of ECT(0) rather than classical marking? .. which might indicate that ECT(0) must not be put in the L4S queue.

**Comment by gwhiteCL** [comment]
One issue with adding ECT(0) in that same sentence is that this means that DSCP 0 + ECT(0) packets (which are expected to be carried in the C queue) would be RECOMMENDED by this statement to be treated the same as DSCP NQB + ECT(0) packets (which are expected to be carried in the L queue) - by *all* functions described in Section 5 of RFC9331.  

Since Section 5 of RFC9331 talks about how the packets in the L queue and the C queue are treated differently from one another, this is a problem.  RFC9331 allows:

> The specification of an L4S AQM MUST define the behaviour for packets with unexpected combinations of codepoints, e.g., a non-ECN-based classifier for the L queue but with ECT(0) in the IP-ECN field (for examples with appropriate behaviours, see Section 2.5.1.1 of the DualQ spec [RFC9332]).

... and that section in RFC9332 recommends:

> if the packet is ECT(0), the L AQM SHOULD apply CE marking using a probability appropriate to Classic congestion control and appropriate to the target delay in the L queue

... which is clearly different than the treatment that an ECT(0) packet in the C queue would receive.

**Comment by gorryfair** [comment]
I thought you would say something like that, so the L4S node ought to place all traffic marked ECT(0) in the L queue [RFC9332]?

**Comment by gwhiteCL** [comment]
Hmm. No, RFC9331 & RFC9332 require that ECT(0) traffic ordinarily goes in the C queue, but they allow that other classifiers might be used to direct traffic to L that is deemed compatible (such as NQB-marked traffic), and they provide requirements and recommendations in case some of that other traffic is ECT(0). 

I don't think we need to write specific requirements around that case in the NQB draft, since the L4S RFCs already are written with NQB in mind when they discuss ECT(0) handling.

**Comment by gorryfair** [comment]
I intended /ought to/ought not to/, however your last point is what matters for resolving this, and so could we just add:

"The processing by an L4S node of a packet that is marked with a NQB DSCP and that also sets an ECT(0) code point is specified in RFC9331 and RFC9332."

**Comment by gwhiteCL** [comment]
Yes, that works.
